board/safety/safety_tesla.h

```diff
@@ -0,0 +1,319 @@
+    if ((controls_allowed == 1) && (eac_status != 0) && (eac_status != 1) && (eac_status != 2))
+    {
+      controls_allowed = 0;
+      puts("EPAS error! \n");
+    }
+  }
+  //get latest steering wheel angle
+  if (addr == 0x00E)
+  {
+    int angle_meas_now = (int)((((to_push->RDLR & 0x3F) << 8) + ((to_push->RDLR >> 8) & 0xFF)) * 0.1 - 819.2);
+    uint32_t ts = TIM2->CNT;
+    uint32_t ts_elapsed = get_ts_elapsed(ts, tesla_ts_angle_last);
+
+    // *** angle real time check
+    // add 1 to not false trigger the violation and multiply by 25 since the check is done every 250 ms and steer angle is updated at     100Hz
+    int rt_delta_angle_up = ((int)((interpolate(TESLA_LOOKUP_ANGLE_RATE_UP, tesla_speed) * 25. + 1.)));
+    int rt_delta_angle_down = ((int)((interpolate(TESLA_LOOKUP_ANGLE_RATE_DOWN, tesla_speed) * 25. + 1.)));
+    int highest_rt_angle = tesla_rt_angle_last + (tesla_rt_angle_last > 0 ? rt_delta_angle_up : rt_delta_angle_down);
+    int lowest_rt_angle = tesla_rt_angle_last - (tesla_rt_angle_last > 0 ? rt_delta_angle_down : rt_delta_angle_up);
+
+    if ((ts_elapsed > TESLA_RT_INTERVAL) || (controls_allowed && !tesla_controls_allowed_last))
+    {
+      tesla_rt_angle_last = angle_meas_now;
+      tesla_ts_angle_last = ts;
+    }
+
+    // update array of samples
+    update_sample(&tesla_angle_meas, angle_meas_now);
+
+    // check for violation;
+    if (max_limit_check(angle_meas_now, highest_rt_angle, lowest_rt_angle))
+    {
+      // We should not be able to STEER under these conditions
+      // Other sending is fine (to allow human override)
+      steer_allowed = 0;
+      puts("WARN: RT Angle - No steer allowed! \n");
+    }
+    else
+    {
+      steer_allowed = 1;
+    }
+
+    tesla_controls_allowed_last = controls_allowed;
+  }
+}
+
+// all commands: gas/regen, friction brake and steering
+// if controls_allowed and no pedals pressed
+//     allow all commands up to limit
+// else
+//     block all commands that produce actuation
+
+static int tesla_tx_hook(CAN_FIFOMailBox_TypeDef *to_send)
+{
+
+  uint32_t addr;
+  int angle_raw;
+  int desired_angle;
+
+  addr = to_send->RIR >> 21;
+
+  // do not transmit CAN message if steering angle too high
+  // DAS_steeringControl::DAS_steeringAngleRequest
+  if (addr == 0x488)
+  {
+    angle_raw = ((to_send->RDLR & 0x7F) << 8) + ((to_send->RDLR & 0xFF00) >> 8);
+    desired_angle = angle_raw * 0.1 - 1638.35;
+    int16_t violation = 0;
+    int st_enabled = (to_send->RDLR & 0x400000) >> 22;
+
+    if (st_enabled == 0) {
+      //steering is not enabled, do not check angles and do send
+      tesla_desired_angle_last = desired_angle;
+      return true;
+    }
+
+    if (controls_allowed)
+    {
+      if (steer_allowed)
+      {
+
+        // add 1 to not false trigger the violation
+        int delta_angle_up = (int)(interpolate(TESLA_LOOKUP_ANGLE_RATE_UP, tesla_speed) * 25. + 1.);
```

Comment:
I think all of these should be floats

Reply:
Yeah, so, we could change these to floats, but then the toyota_ipas safety code is different.  Should we make them the same?

Reply:
yes, I would be happy to generalize those functions and use them in both places

Reply:
If you would be so kind to do that, I would appreciate it, just because I didn't write that code myself and would likely introduce more bugs trying to fix it.

Reply:
yes, once the new functions are defined, I'll take care of using them in safety_toyota_ipas

Reply:
Oh, so you'd like the Tesla team to generalize them and put them in safety.h using floats, or were you going to do that?  Just want to be clear...

Reply:
tesla team shoul'd do that :)